### PR TITLE
[feature] exit gracefully when pressing ctrl+c

### PR DESCRIPTION
**Pull request by Chilledheart**
qt won't register these handlers for us. and when it receives SIGINT,
it exits directly without destroying and releasing objects and memory.

btw, under linux, qt uses glib as its internal event dispatcher, so
calling signal (2) won't work.

### CODE REVIEW

File: src/main.cpp [hunk 1]
```diff
@@ -22,6 +22,44 @@
...
+bool setSignalHandler()
+{
+    struct sigaction sigint_action;
+    sigint_action.sa_handler = sigintHandler;
+    sigemptyset(&sigint_action.sa_mask);
+    sigint_action.sa_flags = 0;
+    sigint_action.sa_flags |= SA_RESTART;
+    if (sigaction(SIGINT, &sigint_action, 0) != 0) // 0 if okay, -1 if failed
+        return false;
+    return true;
```
Comment (lins05) [inline src/main.cpp]: Can we simplify these into `signal(SIGINT, sigintHanlder)`?

Comment (lins05) [inline src/main.cpp]: > btw, under linux, qt uses glib as its internal event dispatcher, so
> calling signal (2) won't work.

I've tried that under ubuntu 14.04, and it seems to work fine.

Comment (Chilledheart) [inline src/main.cpp]: nice, i tried it on os x 10.10 and fedora 21 and it worked as expected.

Comment (Chilledheart) [inline src/main.cpp]: In fact there is a document from qt for it : https://qt-project.org/doc/qt-4.8/unix-signals.html

**Comment by freeplant** [comment]
“exit without destroying and releasing objects and memory”: Why this is matter? After program exit, the memory will be released by the operating system.

**Comment by Chilledheart** [comment]
No, (when sigint is triggerred) it is calling `_exit` which is abnormal exit instead of standard `exit`

**Comment by Chilledheart** [comment]
Here is a small example from [stackoverflow] which you can save to a file `test.cc` and compile it by `g++ -o test test.cc`:
```cpp
#include <iostream>
#include <signal.h>
#include <cstring>
#include <unistd.h>

bool quit = false;    // signal flag

void got_signal(int)
{
    quit = true;
}

class Foo
{
public:
    ~Foo() { std::cout << "destructor\n"; }
};

int main(void)
{
    struct sigaction sa;
    memset( &sa, 0, sizeof(sa) );
    sa.sa_handler = got_signal;
    sigfillset(&sa.sa_mask);
    sigaction(SIGINT,&sa,NULL);

    Foo foo;    // needs destruction before exit
    while (true)
    {
        // do real work here...
        sleep(1);
        if( quit ) break;    // exit normally after SIGINT
    }
    return 0;
}
```

The program calls method `sigaction` to register `SIGINT` signal handler as this PR does. when you press `CTRL+C`, it should print `destructor`, which means the dtor of the c++ object is destroyed normally.

Howevery, if you comment out the `sigaction`, and recompile it, you will find it fails to print anything. It is an abnormal exit now.

see more: 
- [origin post from stackoverflow][stackoverflow]
- http://www.cons.org/cracauer/sigint.html

[stackoverflow]: http://stackoverflow.com/questions/4250013/is-destructor-called-if-sigint-or-sigstp-issued

**Pull request closed without merging**